### PR TITLE
Visit "Any" types in payload visitor

### PR DESCRIPTION
**Pull request by cretz**
**What changed?**

* Added `WellKnownAnyVisitor` to `VisitPayloadsOptions` with default for visiting any
* Updated generator to make sure "Any" types are visited

Closes https://github.com/temporalio/sdk-go/issues/1622

### CODE REVIEW

File: cmd/proxygenerator/interceptor.go [hunk 1]
```diff
@@ -193,15 +215,15 @@ func visitPayloads(
 				if o == nil { continue }
 				no, err := visitPayload(ctx, options, parent, o)
 				if err != nil { return err }
-                objs[i] = no
```
Comment (cretz) [inline cmd/proxygenerator/interceptor.go:196]: This file has some whitespace changes (this was using spaces which was looking bad even though the final generated source was always go fmt'd anyways)

File: cmd/proxygenerator/interceptor.go [hunk 2]
```diff
@@ -159,6 +162,25 @@ func NewFailureVisitorInterceptor(options FailureVisitorInterceptorOptions) (grp
...
 }
 
+func (o *VisitPayloadsOptions) defaultWellKnownAnyVisitor(ctx *VisitPayloadsContext, p *anypb.Any) error {
+	child, err := p.UnmarshalNew()
+	if err != nil {
+		return fmt.Errorf("failed to unmarshal any: %w", err)
+	}
+	// We choose to visit and re-marshal always instead of cloning, visiting,
+	// and checking if anything changed before re-marshaling. It is assumed the
+	// clone + equality check is not much cheaper than re-marshal.
```
Comment (cretz) [inline cmd/proxygenerator/interceptor.go:172]: Does everyone agree with this assumption?

File: cmd/proxygenerator/interceptor.go [hunk 3]
```diff
@@ -159,6 +162,25 @@ func NewFailureVisitorInterceptor(options FailureVisitorInterceptorOptions) (grp
 	}, nil
 }
 
+func (o *VisitPayloadsOptions) defaultWellKnownAnyVisitor(ctx *VisitPayloadsContext, p *anypb.Any) error {
```
Comment (cretz) [inline cmd/proxygenerator/interceptor.go:165]: There could be a need in the future to expose this for users to delegate to, but we will wait until that need to expose it

**Review by Quinn-With-Two-Ns — APPROVED** [review]

File: cmd/proxygenerator/interceptor.go (hunk 2 above)
Comment (tdeebswihart) [reply cmd/proxygenerator/interceptor.go:172]: Can we benchmark that assumption? I'd like to see numbers personally

Comment (cretz) [reply cmd/proxygenerator/interceptor.go:172]: We can get numbers. Can you define the parameters of the benchmarks for your numbers? Do you want the only way we use "any" today (update requests that have barely anything in them) or benchmarking this with more possibilities in "any"? And do you have recommend sizes of payloads? Meaning do you want to benchmark an update with 4K byte inputs or just 5 bytes (or a bunch of other values)?

Also, it should be noted that this utility is mostly used for payload manipulation which almost always mutates the payloads no matter what making this check-never-mutated piece moot.

Comment (tdeebswihart) [reply cmd/proxygenerator/interceptor.go:172]: If that's the context around this assumption, then we should encode that in the code so it's obvious next time looks at an undocumented assumption :)

Comment (cretz) [reply cmd/proxygenerator/interceptor.go:172]: It was not meant to be the context around the assumption. I definitely made the assumption regardless of that context. I can provide benchmark numbers, I just need some context around which numbers/scenarios you want to see.

Comment (cretz) [reply cmd/proxygenerator/interceptor.go:172]: Interesting results here. So here are some basic benchmarks:

```go
func Benchmark_AnyCloneEqualCheck(b *testing.B) {
	anyMsg, _ := anypb.New(&update.Request{Input: &update.Input{Args: &common.Payloads{
		Payloads: []*common.Payload{{Data: []byte("orig-val")}},
	}}})
	for i := 0; i < b.N; i++ {
		child, _ := anyMsg.UnmarshalNew()
		childOrig := proto.Clone(child)
		// Do visitor stuff...
		proto.Equal(childOrig, child)
	}
}

func Benchmark_AnyMarshalReMarshal(b *testing.B) {
	anyMsg, _ := anypb.New(&update.Request{Input: &update.Input{Args: &common.Payloads{
		Payloads: []*common.Payload{{Data: []byte("orig-val")}},
	}}})
	for i := 0; i < b.N; i++ {
		child, err := anyMsg.UnmarshalNew()
		require.NoError(b, err)
		// Do visitor stuff...
		_ = anyMsg.MarshalFrom(child)
	}
}
```

Running this (on Windows at least) shows that the current unmarshal-remarshal way is at least 3x faster every time I try:

```
Benchmark_AnyCloneEqualCheck
Benchmark_AnyCloneEqualCheck-20           464596              2538 ns/op
Benchmark_AnyMarshalReMarshal
Benchmark_AnyMarshalReMarshal-20         1717684               810.9 ns/op
```

**Pull request merged**